src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperContextExtensions.cs

```diff
@@ -28,39 +28,102 @@ public static class TagHelperContextExtensions
+            if (attributes.Missing.Any())
             {
-                if (!context.AllAttributes.ContainsKey(attr)
-                    || context.AllAttributes[attr] == null
-                    || string.IsNullOrWhiteSpace(context.AllAttributes[attr] as string))
+                if (attributes.Present.Any() && logger != null && logger.IsEnabled(LogLevel.Warning))
                 {
-                    // Missing attribute!
-                    missingAttrNames.Add(attr);
+                    // At least 1 attribute was present indicating the user intended to use the tag helper,
+                    // but at least 1 was missing too, so log a warning with the details.
+                    logger.WriteWarning(new MissingAttributeLoggerStructure(context.UniqueId, attributes.Missing));
                 }
-                else
+
+                return false;
+            }
+
+            // All required attributes present
+            return true;
+        }
+
+        /// <summary>
+        /// Determines the mode a <see cref="ITagHelper" /> should run in based on which mode has all its required
+        /// attributes present, non null, non empty, and non whitepsace.
+        /// </summary>
+        /// <typeparam name="TMode">The type representing the <see cref="ITagHelper" />'s modes.</typeparam>
+        /// <typeparam name="TSet">The type representing which attributes are required for which mode.</typeparam>
+        /// <param name="context">The <see cref="TagHelperContext"/>.</param>
+        /// <param name="modeRequiredAttributes">The modes and their required attributes.</param>
+        /// <param name="logger">An <see cref="ILogger"/> to log messages to.</param>
+        /// <returns>The <see cref="ModeMatchResult{TMode}"/>.</returns>
+        public static ModeMatchResult<TMode> DetermineMode<TMode, TSet>(
+            [NotNull] this TagHelperContext context,
+            [NotNull] IEnumerable<Tuple<TMode, TSet>> modeRequiredAttributes,
+            ILogger logger)
+            where TSet : IEnumerable<string>
+        {
+            List<Tuple<TMode, IEnumerable<string>>> modeCandidates = null;
+
+            foreach (var set in modeRequiredAttributes)
+            {
+                var attributes = GetPresentMissingAttributes(context, set.Item2);
+
+                if (!attributes.Missing.Any())
+                {
+                    return ModeMatchResult.Matched(set.Item1);
+                }
+
+                if (attributes.Missing.Any() && attributes.Present.Any())
                 {
-                    atLeastOnePresent = true;
+                    // The set had some present attributes but others missing so capture details of those missing to
+                    // log later on if no match is found
+                    modeCandidates = modeCandidates ?? new List<Tuple<TMode, IEnumerable<string>>>();
+                    modeCandidates.Add(Tuple.Create(set.Item1, attributes.Missing));
```

Comment:
lets use private classes instead of Tuples like @NTaylorMullen correctly noted on the `ScriptTagHelper` PR, it makes the code a whole lot more readable